### PR TITLE
SQL: Limit how much space some string functions can use

### PR DESCRIPTION
**Pull request by bpintea**
This will check and fail if certain functions would generate a result exceeding a certain fixed byte size.

This prevents an operation/query to fail the entire VM.

### CODE REVIEW

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/StringProcessor.java [hunk 1]
```diff
@@ -125,6 +130,17 @@ StringOperation processor() {
         return processor;
     }
 
+    static void checkResultLength(long needed) {
+        if (needed > MAX_RESULT_LENGTH) {
+            throw new SqlIllegalArgumentException(
+                "Required result length [{}] exceeds implementation limit [{}]",
+                needed, // doProcess() below checks that c is within the Integer range
```
Comment (bpintea) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/StringProcessor.java:137]: will remove outdated comment.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/StringProcessor.java [hunk 2]
```diff
@@ -125,6 +130,17 @@ StringOperation processor() {
         return processor;
     }
 
+    static void checkResultLength(long needed) {
+        if (needed > MAX_RESULT_LENGTH) {
+            throw new SqlIllegalArgumentException(
+                "Required result length [{}] exceeds implementation limit [{}]",
```
Comment (astefan) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/string/StringProcessor.java:136]: I would add in the error message the unit of measure (bytes).

**Review by astefan — COMMENTED** [review]
I think this behavior enforcement would be better suited with an interface. Meaning each Processor that needs/must have this should implement.... StringResultLengthLimiter (or whatever) and should provide its own formula for the `long` value to be compared with. Or any other abstraction that enforces this behavior in a more strict manner.

**Comment by bpintea** [comment]
> each Processor that needs/must have this should implement.... StringResultLengthLimiter (or whatever)

That could be done, but the invocation of the checker itself would still need to be "manual", since there's no uniformity across processors about how they deal with input and output (or, more importantly, how they build/allocate the output buffer). Which would not achieve the "stricter" enforcement. But will give it a bit more thought.

**Review by luigidellaquila — APPROVED** [review]
> but the invocation of the checker itself would still need to be "manual" 

I agree with @astefan that it would be good to have a more strict enforcement of this constraint, but I have your same problem: I can't think of a way to make it less "manual".

Apart from that, LGTM

**Review by astefan — APPROVED** [review]
LGTM

**Pull request merged**